src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java

```diff
@@ -263,73 +372,142 @@ public String getRemote() {
-                continue;
-            }
-            if (criteria != null) {
-                SCMSourceCriteria.Probe probe = getProbe(branchName, "pull request", "refs/pull/" + head.getNumber() + "/head", repo, listener);
-                if (criteria.isHead(probe, listener)) {
-                    // FYI https://developer.github.com/v3/pulls/#response-1
-                    Boolean mergeable = ghPullRequest.getMergeable();
-                    if (Boolean.FALSE.equals(mergeable)) {
-                        listener.getLogger().format("      Not mergeable, but it will be included%n");
-                    }
-                    listener.getLogger().format("    Met criteria%n");
-                } else {
-                    listener.getLogger().format("    Does not meet criteria%n");
+                if (!fork && !buildOriginPRMerge && !buildOriginPRHead) {
+                    listener.getLogger().format("    Submitted from origin repository, skipping%n%n");
                     continue;
                 }
+                if (!fork) {
+                    originBranchesWithPR.add(ghPullRequest.getHead().getRef());
+                }
+                boolean trusted = isTrusted(repo, ghPullRequest);
+                if (!trusted) {
+                    listener.getLogger().format("    (not from a trusted source)%n");
+                }
+                for (boolean merge : new boolean[] {false, true}) {
+                    String branchName = "PR-" + number;
+                    if (merge && fork) {
+                        if (!buildForkPRMerge) {
+                            continue; // not doing this combination
+                        }
+                        if (buildForkPRHead) {
+                            branchName += "-merge"; // make sure they are distinct
+                        }
+                        // If we only build merged, or only unmerged, then we use the /PR-\d+/ scheme as before.
+                    }
+                    if (merge && !fork) {
+                        if (!buildOriginPRMerge) {
+                            continue;
+                        }
+                        if (buildForkPRHead) {
+                            branchName += "-merge";
+                        }
+                    }
+                    if (!merge && fork) {
+                        if (!buildForkPRHead) {
+                            continue;
+                        }
+                        if (buildForkPRMerge) {
+                            branchName += "-head";
+                        }
+                    }
+                    if (!merge && !fork) {
+                        if (!buildOriginPRHead) {
+                            continue;
+                        }
+                        if (buildOriginPRMerge) {
+                            branchName += "-head";
+                        }
+                    }
```

Comment:
> Could that go into a `getBranchName` function

Yes, though it is a bit more complicated since we also need to conditionally `continue`.

> A type dedicated to the `build*` props

See above.